subsys/net/lib/tls_credentials/tls_credentials_trusted.c

```diff
@@ -456,3 +460,399 @@ int tls_credential_delete(sec_tag_t tag, enum tls_credential_type type)
+	psa_set_key_bits(&key_attributes, 256);
+
+	psa_status = psa_crypto_init();
+
+	if (psa_status != PSA_SUCCESS) {
+		LOG_ERR("Failed to initialize crypto. Status: %d", psa_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	psa_status = psa_generate_key(&key_attributes, &key_id);
+
+	if (psa_status != PSA_SUCCESS) {
+		LOG_ERR("Failed to generate private key. Status: %d", psa_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* Hand key off to MbedTLS for immediate use (CSR and formatted export). */
+	mbed_status = mbedtls_pk_setup_opaque(&pk_ctx, key_id);
+
+	if (mbed_status) {
+		LOG_ERR("Failed to set up opaque private key. Status: %d", mbed_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* Reset key buffer before using it */
+	memset(key_buf, 0, key_max);
+	*key_len = 0;
+
+	/* Export private key material in the same RFC5915/SEC1 DER format that
+	 * MbedTLS will later expect when loading the private key material from storage.
+	 *
+	 * Temporarily use the public key output buffer for this.
+	 */
+	mbed_status = mbedtls_pk_write_key_der(&pk_ctx, key_buf, key_max);
+	if (mbed_status == MBEDTLS_ERR_ASN1_BUF_TOO_SMALL) {
+		LOG_ERR("Failed to format private key material. Provided buffer too small.");
+		ret = -EFBIG;
+		goto cleanup;
+	}
+	if (mbed_status < 0) {
+		LOG_ERR("Failed to format private key material. Status: %d", mbed_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* On success, mbed_status contains the number of bytes written. */
+
+	/* Place the formatted private key in credentials storage.
+	 * Note that mbedtls_pk_write_key_der writes the key to the end of the key buffer,
+	 * hence the need for key_start.
+	 */
+	key_start = (char*)key_buf + key_max - mbed_status;
+	cred_status = tls_credential_add(tag, TLS_CREDENTIAL_PRIVATE_KEY, key_start, mbed_status);
+	if (cred_status != 0) {
+		LOG_ERR("Error storing private key: %d", cred_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* Clear key buffer afterwards to prevent accidental private key leak */
+	memset(key_buf, 0, key_max);
+
+	/* Finally, write the public key to the output buffer in ASN.1 DER format
+	 * (X.509 SubjectPublicKeyInfo entry. See RFC5280)
+	 */
+	mbed_status = mbedtls_pk_write_pubkey_der(&pk_ctx, key_buf, key_max);
+	if (mbed_status == MBEDTLS_ERR_ASN1_BUF_TOO_SMALL) {
+		LOG_ERR("Failed to format public key. Provided buffer too small.");
+		ret = -EFBIG;
+		goto cleanup;
+	}
+	if (mbed_status < 0) {
+		LOG_ERR("Failed to format public key. Status: %d", mbed_status);
+		ret = -EFAULT;
+		goto cleanup;
+	}
+
+	/* On success, mbed_status contains the number of bytes written. */
+	*key_len = mbed_status;
+
+	/* mbedtls_pk_write_pubkey_der writes its data to the end of the key buffer.
+	 * Shift this data to the beginning of the key buffer for convenient use by the caller.
+	 */
+	key_start = (char*)key_buf + key_max - mbed_status;
+	memmove(key_buf, key_start, mbed_status);
+
+cleanup:
+	k_mutex_unlock(&credential_lock);
```

Comment:
why is `credentials_unlock()` not used here?  only used when called from the shell?